installinator/src/write.rs

```diff
@@ -448,32 +452,80 @@ impl ArtifactsToWrite<'_> {
         transport: &mut impl WriteTransport,
         cx: &StepContext<WriteSpec>,
     ) -> Result<StepResult<(), WriteSpec>, WriteError> {
-        // Temporary workaround while we may not know how to write the control
-        // plane image: if we don't know where to put it, we're done.
-        let Some(control_plane_dest) = destinations.control_plane.as_ref() else
-        {
-            return StepResult::skipped((), (), "We don't yet know how to write the control plane");
-        };
-
-        write_artifact_impl(
-            WriteComponent::ControlPlane,
+        // Register a nested engine to write the set of zones, each zone as its
+        // own step.
+        let inner_cx = ControlPlaneZoneWriteContext {
             slot,
-            self.control_plane_data.clone(),
-            control_plane_dest,
-            true,
-            transport,
-            cx,
-        )
+            output_directory: &destinations.control_plane_dir,
+            zones: self.control_plane_zones,
+        };
+        cx.with_nested_engine(|engine| {
+            inner_cx.register_steps(engine, transport);
+            Ok(())
+        })
         .await
         .map_err(|error| {
-            info!(log, "{error:?}"; "artifact_id" => ?self.control_plane_id);
+            warn!(
+                log, "{error:?}";
+                "artifact_id" => ?self.control_plane_id,
+            );
             error
         })?;
 
+        info!(
+            log,
+            "finished writing {} control plane zones",
+            self.control_plane_zones.zones.len()
+        );
         StepResult::success((), ())
     }
 }
 
+struct ControlPlaneZoneWriteContext<'a> {
+    slot: M2Slot,
+    output_directory: &'a Utf8Path,
+    zones: &'a ControlPlaneZoneImages,
+}
+
+impl ControlPlaneZoneWriteContext<'_> {
+    fn register_steps<'b>(
+        &'b self,
+        engine: &UpdateEngine<'b, ControlPlaneZonesSpec>,
+        transport: &'b mut impl WriteTransport,
+    ) {
+        use update_engine::StepHandle;
+
+        let slot = self.slot;
+        let mut transport = StepHandle::ready(transport);
```

Comment:
I don't think you need `StepHandle::ready` in this case (it's mainly useful for when you have conditional steps -- a step returns a value in some cases and not in others).

Reply:
Ah, I should put a comment here. I need to do this because I can't share the `&mut impl WriteTransport` we get as an argument with every step, so instead I stash it into a `StepHandle` and reassign it on every loop iteration: each step passes it on to the next.

Reply:
Added an explanation in 18175672

Reply:
Ah understood!